docs/content/concepts/configuration/advanced-config-types.mdx

```diff
@@ -0,0 +1,276 @@
+
+Here, we define a config schema which takes in a `pet` field, which can be either a `Cat` or a `Dog`, as indicated by the `pet_type` field.
+
+```python file=/guides/dagster/pythonic_config/pythonic_config.py startafter=start_union_schema_config endbefore=end_union_schema_config dedent=4
+from dagster import asset, materialize, Config, RunConfig
+from pydantic import Field
+from typing import Union
+from typing_extensions import Literal
+
+class Cat(Config):
+    pet_type: Literal["cat"] = "cat"
+    meows: int
+
+class Dog(Config):
+    pet_type: Literal["dog"] = "dog"
+    barks: float
+
+class ConfigWithUnion(Config):
+    # Here, the ellipses `...` indicates that the field is required and has no default value.
+    pet: Union[Cat, Dog] = Field(..., discriminator="pet_type")
+
+@asset
+def pet_stats(config: ConfigWithUnion):
+    if isinstance(config.pet, Cat):
+        return f"Cat meows {config.pet.meows} times"
+    else:
+        return f"Dog barks {config.pet.barks} times"
+
+result = materialize(
+    [pet_stats],
+    run_config=RunConfig(
+        {
+            "pet_stats": ConfigWithUnion(
+                pet=Cat(meows=10),
+            )
+        }
+    ),
+)
+```
+
+---
+
+## Enum types
+
+Python enums which subclass `Enum` are supported as config fields. Here, we define a schema that takes in a list of users, whose roles are specified as enum values:
+
+```python file=/guides/dagster/pythonic_config/pythonic_config.py startafter=start_enum_schema_config endbefore=end_enum_schema_config dedent=4
+from dagster import Config, RunConfig, op, job
+from enum import Enum
+
+class UserPermissions(Enum):
+    GUEST = "guest"
+    MEMBER = "member"
+    ADMIN = "admin"
+
+class ProcessUsersConfig(Config):
+    users_list: Dict[str, UserPermissions]
+
+@op
+def process_users(config: ProcessUsersConfig):
+    for user, permission in config.users_list.items():
+        if permission == UserPermissions.ADMIN:
+            print(f"{user} is an admin")  # noqa: T201
+
+@job
+def process_users_job():
+    process_users()
+
+op_result = process_users_job.execute_in_process(
+    run_config=RunConfig(
+        {
+            "process_users": ProcessUsersConfig(
+                users_list={
+                    "Bob": UserPermissions.GUEST,
+                    "Alice": UserPermissions.ADMIN,
+                }
+            )
+        }
+    ),
+)
+```
+
+---
+
+## Validated config fields
+
+Config fields can have custom validation logic applied using [Pydantic validators](https://docs.pydantic.dev/usage/validators/). Pydantic validators are defined as methods on the config class, and are decorated with the `@validator` decorator. Here, we define some validators on a configured user's name and username, which will throw exceptions if incorrect values are passed in the launchpad or from a schedule or sensor.
```

Comment:
Worth detailing _exactly_ when these is evaluated. For example a failing validator will still allow the launchpad launch button to be pressed and will result in a runtime error, but will prevent launch.